### ISSUE TITLE
E1019: Parameter for Fn::Sub not found at Resources

### ISSUE THREAD

**Issue by Trandel** [aws-cloudformation/cfn-python-lint#957]
*cfn-lint version: (`0.21.4`)*

Linting is creating false errors about Parameters not found for a Sub function.

It looks like only version `0.21.4` is affected. Version `0.21.3` works as expected.

The part of the template that throws the error is:

```yaml
AWSTemplateFormatVersion: 2010-09-09

Description: ECSCC Serverless Application for fetching the status of pipelines

Transform: AWS::Serverless-2016-10-31

Parameters:

  pFunctionAliasName:
    Description: The Lambda Function Alias name.
    Type: String
    Default: Live

  pArtifactS3BucketName:
    AllowedPattern: ^[0-9a-zA-Z]+([0-9a-zA-Z-.]*[0-9a-zA-Z])*$
    ConstraintDescription: bucket name can include numbers, lowercase
      letters, uppercase letters, periods (.), and hyphens (-). It cannot start or
      end with a hyphen (-).
    Description: S3 bucket name for the Cfn assets. bucket name
      can include numbers, lowercase letters, uppercase letters, and hyphens (-).
      It cannot start or end with a hyphen (-).
    Type: String

  pArtifactS3Key:
    AllowedPattern: ^[0-9a-zA-Z-\/\.]*$
    ConstraintDescription: key can include numbers, lowercase letters,
      uppercase letters, hyphens (-), and forward slash (/). It cannot start or end
      with forward slash (/) because they are automatically appended.
    Description: S3 key for the Lambda zipfile. key
      can include numbers, lowercase letters, uppercase letters, hyphens (-), and
      forward slash (/). It cannot start or end with forward slash (/) because they
      are automatically appended.
    Type: String

  pDeploymentPreference:
    Description: The traffic shaping deployment preference.
    Type: String
    Default: Canary10Percent5Minutes
    AllowedValues:
      - AllAtOnce
      - Linear10PercentEvery10Minutes
      - Linear10PercentEvery1Minute
      - Linear10PercentEvery2Minutes
      - Linear10PercentEvery3Minutes
      - Canary10Percent30Minutes
      - Canary10Percent5Minutes
      - Canary10Percent10Minutes
      - Canary10Percent15Minutes

  pLogLevel:
    Description: verbosity level of function loggers
    Type: String
    Default: DEBUG
    AllowedValues:
      - CRITICAL
      - ERROR
      - WARNING
      - INFO
      - DEBUG

Globals:

  Function:
    Runtime: python3.7
    Timeout: 300

Resources:

  rLambdaCodeDeployRole:
    Type: AWS::IAM::Role
    Metadata:
      cfn_nag:
        rules_to_suppress:
          - id: W11
            reason: "This role should allow any cloudwatch:DescribeAlarms lambda:UpdateAlias lambda:GetAlias sns:Publish."
    Properties:
      AssumeRolePolicyDocument:
        Version: '2012-10-17'
        Statement:
          - Effect: Allow
            Principal:
              Service:
                - codedeploy.amazonaws.com
            Action:
              - sts:AssumeRole
      Path: /
      Policies:
        - PolicyName: lambda-codedeploy
          PolicyDocument:
            Version: '2012-10-17'
            Statement:
              - Action:
                  - cloudwatch:DescribeAlarms
                  - lambda:UpdateAlias
                  - lambda:GetAlias
                  - sns:Publish
                Resource: '*'
                Effect: Allow
              - Action:
                  - s3:GetObject
                  - s3:GetObjectVersion
                Resource: arn:aws:s3:::*/CodeDeploy/*
                Effect: Allow
              - Action:
                  - s3:GetObject
                  - s3:GetObjectVersion
                Resource: '*'
                Condition:
                  StringEquals:
                    s3:ExistingObjectTag/UseWithCodeDeploy: 'true'
                Effect: Allow
              - Action:
                  - lambda:InvokeFunction
                Resource:
                  - !GetAtt rPreTrafficLambdaFunction.Arn
                  - !GetAtt rPostTrafficLambdaFunction.Arn
                Effect: Allow

  rFetchPipelineInfoRole:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Version: '2012-10-17'
        Statement:
          - Effect: Allow
            Principal:
              Service:
                - lambda.amazonaws.com
            Action:
              - sts:AssumeRole
      Path: /
      ManagedPolicyArns:
        - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
        - arn:aws:iam::aws:policy/AWSXrayWriteOnlyAccess

  rFetchPipelineInfo:
    Type: AWS::Serverless::Function
    Properties:
      Handler: fetch_pipeline_info.handler
      Environment:
        Variables:
          LOG_LEVEL: !Ref pLogLevel
      CodeUri:
        Bucket: !Ref pArtifactS3BucketName
        Key: !Ref pArtifactS3Key
      Role: !GetAtt rFetchPipelineInfoRole.Arn
      AutoPublishAlias: !Ref pFunctionAliasName
      DeploymentPreference:
        Type: !Ref pDeploymentPreference
        Role: !GetAtt rLambdaCodeDeployRole.Arn
        Alarms:
          - !Ref rFetchPipelineInfoAliasErrorAlarm
          - !Ref rFetchPipelineInfoVersionErrorAlarm
        Hooks:
          PreTraffic: !Ref rPreTrafficLambdaFunction
          PostTraffic: !Ref rPostTrafficLambdaFunction

  rFetchPipelineInfoAliasErrorAlarm:
    Type: AWS::CloudWatch::Alarm
    Properties:
      AlarmDescription: Lambda Function Error > 0
      ComparisonOperator: GreaterThanThreshold
      Dimensions:
        - Name: Resource
          Value: !Sub ${rFetchPipelineInfo}:${pFunctionAliasName}
        - Name: FunctionName
          Value: !Ref rFetchPipelineInfo
      EvaluationPeriods: 1
      MetricName: Errors
      Namespace: AWS/Lambda
      Period: 300
      Statistic: Sum
      Threshold: 0

  rFetchPipelineInfoVersionErrorAlarm:
    Type: AWS::CloudWatch::Alarm
    Properties:
      AlarmDescription: Lambda Function Error > 0
      ComparisonOperator: GreaterThanThreshold
      Dimensions:
        - Name: Resource
          Value: !Ref rFetchPipelineInfo.Version
        - Name: FunctionName
          Value: !Ref rFetchPipelineInfo
      EvaluationPeriods: 1
      MetricName: Errors
      Namespace: AWS/Lambda
      Period: 300
      Statistic: Sum
      Threshold: 0

  rLambdaPreHookFunctionRole:
    Type: AWS::IAM::Role
    Metadata:
      cfn_nag:
        rules_to_suppress:
          - id: W11
            reason: "This role should allow codedeploy:PutLifecycleEventHookExecutionStatus."
    Properties:
      AssumeRolePolicyDocument:
        Version: '2012-10-17'
        Statement:
          - Effect: Allow
            Principal:
              Service:
                - lambda.amazonaws.com
            Action:
              - sts:AssumeRole
      Policies:
        - PolicyName: lambda-hook
          PolicyDocument:
            Version: '2012-10-17'
            Statement:
              - Effect: Allow
                Action:
                  - codedeploy:PutLifecycleEventHookExecutionStatus
                Resource:
                  !Sub 'arn:aws:codedeploy:${AWS::Region}:${AWS::AccountId}:deploymentgroup:${ServerlessDeploymentApplication}/*'
              - Effect: Allow
                Action:
                  - lambda:InvokeFunction
                Resource: !Ref rFetchPipelineInfo.Version
      Path: /
      ManagedPolicyArns:
        - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
        - arn:aws:iam::aws:policy/AWSXrayWriteOnlyAccess

  rPreTrafficLambdaFunction:
    Type: AWS::Serverless::Function
    Properties:
      Handler: pre_traffic.handler
      CodeUri:
        Bucket: !Ref pArtifactS3BucketName
        Key: !Ref pArtifactS3Key
      Role: !GetAtt rLambdaPreHookFunctionRole.Arn
      Environment:
        Variables:
          FUNCTION_NAME: !Ref rFetchPipelineInfo
          FUNCTION_VERSION: !Ref rFetchPipelineInfo.Version
          LOG_LEVEL: DEBUG
      DeploymentPreference:
        Enabled: false
      Timeout: 5

  rLambdaPostHookFunctionRole:
    Type: AWS::IAM::Role
    Metadata:
      cfn_nag:
        rules_to_suppress:
          - id: W11
            reason: "This role should allow codedeploy:PutLifecycleEventHookExecutionStatus."
    Properties:
      AssumeRolePolicyDocument:
        Version: '2012-10-17'
        Statement:
          - Effect: Allow
            Principal:
              Service:
                - lambda.amazonaws.com
            Action:
              - sts:AssumeRole
      Policies:
        - PolicyName: lambda-hook
          PolicyDocument:
            Version: '2012-10-17'
            Statement:
              - Effect: Allow
                Action:
                  - codedeploy:PutLifecycleEventHookExecutionStatus
                Resource:
                  !Sub 'arn:aws:codedeploy:${AWS::Region}:${AWS::AccountId}:deploymentgroup:${ServerlessDeploymentApplication}/*'
              - Effect: Allow
                Action:
                  - lambda:InvokeFunction
                Resource: !Ref rFetchPipelineInfo.Version
      Path: /
      ManagedPolicyArns:
        - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
        - arn:aws:iam::aws:policy/AWSXrayWriteOnlyAccess

  rPostTrafficLambdaFunction:
    Type: AWS::Serverless::Function
    Properties:
      Handler: post_traffic.handler
      CodeUri:
        Bucket: !Ref pArtifactS3BucketName
        Key: !Ref pArtifactS3Key
      Role: !GetAtt rLambdaPostHookFunctionRole.Arn
      Environment:
        Variables:
          FUNCTION_NAME: !Ref rFetchPipelineInfo
          FUNCTION_VERSION: !Ref rFetchPipelineInfo.Version
          LOG_LEVEL: DEBUG
      DeploymentPreference:
        Enabled: false
      Timeout: 5

```

I'm running lint like this:

`cfn-lint -f parseable -c I -t template.yml`

and getting the following error:

`template.yml:165:11:165:16:E1019:Parameter pFunctionAliasName for Fn::Sub not found at Resources/rFetchPipelineInfoAliasErrorAlarm/Properties/Dimensions/0/Value/Fn::Sub
`

**Comment by kddejong** [comment]
#947 should fix this.  We will have it released today.

**Comment by kddejong** [comment]
Should be fixed with v0.21.5.  Let me know if it doesn't work still.
